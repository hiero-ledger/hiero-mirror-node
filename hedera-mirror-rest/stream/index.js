// SPDX-License-Identifier: Apache-2.0

import CompositeRecordFile from './compositeRecordFile';
import SignatureFile from './signatureFile';

export {CompositeRecordFile, SignatureFile};
