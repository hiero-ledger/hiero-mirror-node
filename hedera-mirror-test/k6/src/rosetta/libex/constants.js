// SPDX-License-Identifier: Apache-2.0

export const currencyHbar = {
  symbol: 'HBAR',
  decimals: 8,
  metadata: {
    issuer: 'Hedera',
  },
};
