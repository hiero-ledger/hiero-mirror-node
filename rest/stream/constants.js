// SPDX-License-Identifier: Apache-2.0

const BYTE_SIZE = 1;
const INT_SIZE = 4;
const LONG_SIZE = 8;

export {BYTE_SIZE, INT_SIZE, LONG_SIZE};
