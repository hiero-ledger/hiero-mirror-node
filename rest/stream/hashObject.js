// SPDX-License-Identifier: Apache-2.0

import {INT_SIZE} from './constants';
import StreamObject from './streamObject';
import {readLengthAndBytes} from './utils';

class HashObject extends StreamObject {
  // properties of SHA-384 hash algorithm
  static SHA_384 = {
    encoding: 'hex',
    length: 48,
    name: 'sha384',
  };

  /**
   * Reads the body of the hash object
   * @param {Buffer} buffer
   * @returns {Number} The size of the body in bytes
   */
  _readBody(buffer) {
    // always SHA-384
    const hashLength = HashObject.SHA_384.length;
    this.digestType = buffer.readInt32BE();
    const {length, bytes} = readLengthAndBytes(buffer.subarray(INT_SIZE), hashLength, hashLength, false);
    this.hash = bytes;

    return INT_SIZE + length;
  }
}

export default HashObject;
