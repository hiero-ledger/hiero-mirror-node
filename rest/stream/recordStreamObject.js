// SPDX-License-Identifier: Apache-2.0

import {BYTE_SIZE} from './constants';
import StreamObject from './streamObject';
import {readLengthAndBytes} from './utils';

class RecordStreamObject extends StreamObject {
  static MAX_RECORD_LENGTH = 64 * 1024;
  static MAX_TRANSACTION_LENGTH = 64 * 1024;

  /**
   * Reads the body of the record stream object
   * @param {Buffer} buffer
   * @returns {Number} The size of the body in bytes
   */
  _readBody(buffer) {
    const record = readLengthAndBytes(buffer, BYTE_SIZE, RecordStreamObject.MAX_RECORD_LENGTH, false);
    const transaction = readLengthAndBytes(
      buffer.subarray(record.length),
      BYTE_SIZE,
      RecordStreamObject.MAX_TRANSACTION_LENGTH,
      false
    );
    this.record = record.bytes;
    this.transaction = transaction.bytes;

    return record.length + transaction.length;
  }
}

export default RecordStreamObject;
