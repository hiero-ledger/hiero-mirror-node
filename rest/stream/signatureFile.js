// SPDX-License-Identifier: Apache-2.0

import {proto} from '@hashgraph/proto';

import {BYTE_SIZE, INT_SIZE} from './constants';
import HashObject from './hashObject';
import SignatureObject from './signatureObject';
import {readLengthAndBytes, readNBytes} from './utils';

// version, object stream signature version
const V5_FILE_HASH_OFFSET = BYTE_SIZE + INT_SIZE;

class SignatureFile {
  /**
   * Parses signature file buffer, retrieves hashes and the corresponding signatures.
   * @param {Buffer} buffer
   */
  constructor(buffer) {
    const version = buffer.readInt8();
    switch (version) {
      case 4:
        this.version = 2;
        this._parseV2SignatureFile(buffer);
        break;
      case 5:
        this.version = 5;
        this._parseV5SignatureFile(buffer);
        break;
      case 6:
        this.version = 6;
        this._parseV6SignatureFile(buffer);
        break;
      default:
        throw new Error(`Unexpected signature file version '${version}'`);
    }
  }

  _parseV2SignatureFile(buffer) {
    // skip type, already checked
    buffer = buffer.subarray(BYTE_SIZE);
    this.fileHash = readNBytes(buffer, HashObject.SHA_384.length);

    buffer = buffer.subarray(HashObject.SHA_384.length);
    const type = buffer.readInt8();
    if (type !== 3) {
      throw new Error(`Unexpected type delimiter '${type}' in signature file`);
    }

    buffer = buffer.subarray(BYTE_SIZE);
    const {length, bytes} = readLengthAndBytes(buffer, BYTE_SIZE, SignatureObject.SHA_384_WITH_RSA.maxLength, false);
    this.fileHashSignature = bytes;

    buffer = buffer.subarray(length);
    if (buffer.length !== 0) {
      throw new Error('Extra data discovered in signature file ');
    }
  }

  _parseV5SignatureFile(buffer) {
    buffer = buffer.subarray(V5_FILE_HASH_OFFSET);
    const fileHashObject = new HashObject(buffer);
    const fileHashSignatureObject = new SignatureObject(buffer.subarray(fileHashObject.getLength()));

    buffer = buffer.subarray(fileHashObject.getLength() + fileHashSignatureObject.getLength());
    const metadataHashObject = new HashObject(buffer);
    const metadataHashSignatureObject = new SignatureObject(buffer.subarray(metadataHashObject.getLength()));

    if (buffer.length !== metadataHashObject.getLength() + metadataHashSignatureObject.getLength()) {
      throw new Error('Extra data discovered in signature file');
    }

    this.fileHash = fileHashObject.hash;
    this.fileHashSignature = fileHashSignatureObject.signature;
    this.metadataHash = metadataHashObject.hash;
    this.metadataHashSignature = metadataHashSignatureObject.signature;
  }

  _parseV6SignatureFile(buffer) {
    const signatureFile = proto.SignatureFile.decode(buffer.subarray(1));
    this.fileHash = signatureFile.fileSignature.hashObject.hash;
    this.fileHashSignature = signatureFile.fileSignature.signature;
    this.metadataHash = signatureFile.metadataSignature.hashObject.hash;
    this.metadataHashSignature = signatureFile.metadataSignature.signature;
  }
}

export default SignatureFile;
