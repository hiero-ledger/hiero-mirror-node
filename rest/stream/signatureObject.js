// SPDX-License-Identifier: Apache-2.0

import {BYTE_SIZE, INT_SIZE} from './constants';
import StreamObject from './streamObject';
import {readLengthAndBytes} from './utils';

class SignatureObject extends StreamObject {
  // properties of SHA384WithRsa signature
  static SHA_384_WITH_RSA = {
    type: 1,
    maxLength: 384,
  };

  /**
   * Reads the body of the signature object
   * @param {Buffer} buffer
   * @returns {Number} The size of the body in bytes
   */
  _readBody(buffer) {
    const message = 'Error reading signature object';
    this.type = buffer.readInt32BE();
    if (this.type !== SignatureObject.SHA_384_WITH_RSA.type) {
      throw new Error(`${message}, expect type ${SignatureObject.SHA_384_WITH_RSA.type} got ${this.type}`);
    }

    const {length, bytes} = readLengthAndBytes(
      buffer.subarray(INT_SIZE),
      BYTE_SIZE,
      SignatureObject.SHA_384_WITH_RSA.maxLength,
      true
    );
    this.signature = bytes;

    return INT_SIZE + length;
  }
}

export default SignatureObject;
