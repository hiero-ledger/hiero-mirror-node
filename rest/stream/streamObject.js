// SPDX-License-Identifier: Apache-2.0

import {INT_SIZE, LONG_SIZE} from './constants';

// classId, classVersion
const STREAM_OBJECT_HEADER_SIZE = LONG_SIZE + INT_SIZE;

class StreamObject {
  /**
   * Reads stream object from buffer
   * @param {Buffer} buffer - The buffer to read the stream object from
   */
  constructor(buffer) {
    this.classId = buffer.readBigInt64BE();
    this.classVersion = buffer.readInt32BE(LONG_SIZE);

    this.bodyLength = this._readBody(buffer.subarray(STREAM_OBJECT_HEADER_SIZE));
  }

  _readBody(buffer) {
    return 0;
  }

  /**
   * Gets the serialized header with fields in little-endian order.
   *
   * @return {Buffer}
   */
  getHeaderLE() {
    const header = Buffer.alloc(STREAM_OBJECT_HEADER_SIZE);
    header.writeBigInt64LE(this.classId);
    header.writeInt32LE(this.classVersion, LONG_SIZE);
    return header;
  }

  getHeaderLength() {
    return STREAM_OBJECT_HEADER_SIZE;
  }

  getLength() {
    return STREAM_OBJECT_HEADER_SIZE + this.bodyLength;
  }
}

export default StreamObject;
