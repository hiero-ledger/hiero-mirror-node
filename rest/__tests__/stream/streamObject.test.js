// SPDX-License-Identifier: Apache-2.0

import {INT_SIZE, LONG_SIZE} from '../../stream/constants';
import StreamObject from '../../stream/streamObject';

describe('StreamObject', () => {
  let buffer;
  const classId = BigInt.asIntN(64, BigInt('0xdeadbeaf12345678'));
  const classVersion = 0x12345678;
  const header = Buffer.concat([
    Buffer.from([0x78, 0x56, 0x34, 0x12, 0xaf, 0xbe, 0xad, 0xde]),
    Buffer.from([0x78, 0x56, 0x34, 0x12]),
  ]);

  beforeEach(() => {
    buffer = Buffer.alloc(LONG_SIZE + INT_SIZE);
    buffer.writeBigInt64BE(classId);
    buffer.writeInt32BE(classVersion, LONG_SIZE);
  });

  test('getLength', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.getHeaderLength()).toEqual(LONG_SIZE + INT_SIZE);
  });

  test('getLength', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.getLength()).toEqual(LONG_SIZE + INT_SIZE);
  });

  test('getHeaderLE', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.getHeaderLE()).toEqual(header);
  });

  test('classId', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.classId).toEqual(classId);
  });

  test('classVersion', () => {
    const streamObject = new StreamObject(buffer);
    expect(streamObject.classVersion).toEqual(classVersion);
  });

  test('create from empty buffer', () => {
    expect(() => new StreamObject(Buffer.from([]))).toThrowErrorMatchingSnapshot();
  });

  test('create from truncated buffer', () => {
    expect(() => new StreamObject(Buffer.from([1, 2, 3, 4, 5]))).toThrowErrorMatchingSnapshot();
  });
});
