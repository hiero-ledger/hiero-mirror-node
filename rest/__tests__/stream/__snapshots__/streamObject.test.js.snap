// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StreamObject create from empty buffer 1`] = `"Attempt to access memory outside buffer bounds"`;

exports[`StreamObject create from truncated buffer 1`] = `"Attempt to access memory outside buffer bounds"`;
